Sources/NIOWebSocket/WebSocketFrame.swift

```diff
@@ -0,0 +1,269 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+//
+//  WebSocketFrame.swift
+//  NIOWebSocket
+//
+//  Created by Cory Benfield on 16/02/2018.
+//
+
+import NIO
+
+private extension UInt8 {
+    func bitSetWithMask(_ mask: UInt8) -> Bool {
+        return self & mask != 0
+    }
+
+    mutating func setBitForMask(_ mask: UInt8, to: Bool) {
+        if to {
+            self |= mask
+        } else {
+            self &= ~mask
+        }
+    }
+}
+
+/// A single 4-byte websocket masking key.
+///
+/// WebSockets uses a masking key to prevent malicious users from injecting
+/// predictable binary sequences into websocket data streams. This structure provides
+/// a more convenient method of interacting with a masking key than simply by passing
+/// around a four-tuple.
+public struct WebSocketMaskingKey {
+    private let key: (UInt8, UInt8, UInt8, UInt8)
+
+    public init?<T: Collection>(_ buffer: T) where T.Element == UInt8 {
+        guard buffer.count == 4 else {
+            return nil
+        }
+
+
+        self.key = (buffer[buffer.startIndex],
+                    buffer[buffer.index(buffer.startIndex, offsetBy: 1)],
+                    buffer[buffer.index(buffer.startIndex, offsetBy: 2)],
+                    buffer[buffer.index(buffer.startIndex, offsetBy: 3)])
+    }
+}
+
+extension WebSocketMaskingKey: ExpressibleByArrayLiteral {
+    public typealias ArrayLiteralElement = UInt8
+
+    public init(arrayLiteral elements: UInt8...) {
```

Comment:
hmm, this is _not_ an array literal here but the name is `arrayLiteral`. These are varargs. And do we actually need varargs? I don't really like them

Reply:
This *is* an array literal, this is part of the conformance for `ExpressibleByArrayLiteral`. This allows the Swift compiler to do: `maskingKey = [1, 2, 3, 4]`.

Reply:
ok, that is weird. Why is the type then not `[UInt8]`?

Reply:
[Take it up with the stdlib team](https://developer.apple.com/documentation/swift/expressiblebyarrayliteral) ;)